File: 2022/day04/php/solution.php
#!/usr/bin/env php
<?php

function parseInput(string $filename): array {
    $pairs = [];
    $lines = file($filename, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);

    foreach ($lines as $line) {
        $parts = explode(',', $line);
        $left = explode('-', $parts[0]);
        $right = explode('-', $parts[1]);
        $pairs[] = [
            (int)$left[0],
            (int)$left[1],
            (int)$right[0],
            (int)$right[1]
        ];
    }

    return $pairs;
}

function fullyContains(int $a1, int $b1, int $a2, int $b2): bool {
    // Check if one range fully contains the other
    return ($a1 <= $a2 && $b1 >= $b2) || ($a2 <= $a1 && $b2 >= $b1);
}

function overlaps(int $a1, int $b1, int $a2, int $b2): bool {
    // Check if ranges overlap at all
    return $a1 <= $b2 && $a2 <= $b1;
}

function part1(array $pairs): int {
    $count = 0;
    foreach ($pairs as $pair) {
        if (fullyContains($pair[0], $pair[1], $pair[2], $pair[3])) {
            $count++;
        }
    }
    return $count;
}

function part2(array $pairs): int {
    $count = 0;
    foreach ($pairs as $pair) {
        if (overlaps($pair[0], $pair[1], $pair[2], $pair[3])) {
            $count++;
        }
    }
    return $count;
}

function main(): void {
    $inputFile = __DIR__ . '/../input.txt';
    $pairs = parseInput($inputFile);

    echo 'Part 1: ' . part1($pairs) . PHP_EOL;
    echo 'Part 2: ' . part2($pairs) . PHP_EOL;
}

main();

File: 2022/day04/rust/solution.rs
use std::fs;
use std::path::Path;

fn parse_input(filename: &Path) -> Vec<(u32, u32, u32, u32)> {
    let content = fs::read_to_string(filename).expect("Failed to read input file");
    content
        .lines()
        .filter(|line| !line.is_empty())
        .map(|line| {
            let parts: Vec<&str> = line.split(',').collect();
            let left: Vec<u32> = parts[0].split('-').map(|x| x.parse().unwrap()).collect();
            let right: Vec<u32> = parts[1].split('-').map(|x| x.parse().unwrap()).collect();
            (left[0], left[1], right[0], right[1])
        })
        .collect()
}

fn fully_contains(a1: u32, b1: u32, a2: u32, b2: u32) -> bool {
    // Check if one range fully contains the other
    (a1 <= a2 && b1 >= b2) || (a2 <= a1 && b2 >= b1)
}

fn overlaps(a1: u32, b1: u32, a2: u32, b2: u32) -> bool {
    // Check if ranges overlap at all
    a1 <= b2 && a2 <= b1
}

fn part1(pairs: &[(u32, u32, u32, u32)]) -> usize {
    pairs
        .iter()
        .filter(|&&(a1, b1, a2, b2)| fully_contains(a1, b1, a2, b2))
        .count()
}

fn part2(pairs: &[(u32, u32, u32, u32)]) -> usize {
    pairs
        .iter()
        .filter(|&&(a1, b1, a2, b2)| overlaps(a1, b1, a2, b2))
        .count()
}

fn main() {
    let exe_path = std::env::current_exe().expect("Failed to get executable path");
    let exe_dir = exe_path.parent().expect("Failed to get executable directory");
    let input_path = exe_dir.join("../input.txt");

    // Try executable-relative path first, fallback to cwd-relative
    let input_file = if input_path.exists() {
        input_path
    } else {
        Path::new("../input.txt").to_path_buf()
    };

    let pairs = parse_input(&input_file);

    println!("Part 1: {}", part1(&pairs));
    println!("Part 2: {}", part2(&pairs));
}
